File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationBaseFormFragment.java
package org.wordpress.android.ui.sitecreation;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;

import org.wordpress.android.R;

public abstract class SiteCreationBaseFormFragment extends Fragment {
    public static final String EXTRA_SCREEN_TITLE = "extra_screen_title";

    protected abstract @LayoutRes
    int getContentLayout();

    protected abstract void setupContent(ViewGroup rootView);

    protected abstract void onHelp();

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setHasOptionsMenu(true);
    }

    protected ViewGroup createMainView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        ViewGroup rootView = (ViewGroup) inflater.inflate(R.layout.site_creation_form_screen, container, false);
        ViewStub formContainer = (rootView.findViewById(R.id.site_creation_form_content_stub));
        formContainer.setLayoutResource(getContentLayout());
        formContainer.inflate();
        return rootView;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        ViewGroup rootView = createMainView(inflater, container, savedInstanceState);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setupContent((ViewGroup) getView().getRootView());

        Toolbar toolbar = view.findViewById(R.id.toolbar_main);
        ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);

        ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle(getScreenTitle());
            // important for accessibility
            getActivity().setTitle(getScreenTitle());
        }
    }

    protected abstract String getScreenTitle();

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_site_creation, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.help) {
            onHelp();
            return true;
        }

        return false;
    }
}

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationBaseFormFragment.kt
package org.wordpress.android.ui.sitecreation

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import org.wordpress.android.R
import org.wordpress.android.databinding.SiteCreationFormScreenBinding

abstract class SiteCreationBaseFormFragment : Fragment(R.layout.site_creation_form_screen), MenuProvider {
    @LayoutRes protected abstract fun getContentLayout(): Int

    protected abstract val screenTitle: String

    protected abstract fun setupContent()

    protected abstract fun onHelp()

    protected abstract fun setBindingViewStubListener(parentBinding: SiteCreationFormScreenBinding)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requireActivity().addMenuProvider(this, viewLifecycleOwner)

        with(SiteCreationFormScreenBinding.bind(view)) {
            siteCreationFormContentStub.layoutResource = getContentLayout()
            setBindingViewStubListener(this)
            siteCreationFormContentStub.inflate()

            setupContent()

            (requireActivity() as AppCompatActivity).setSupportActionBar(toolbarMain)
            val actionBar = (activity as AppCompatActivity?)!!.supportActionBar
            if (actionBar != null) {
                actionBar.setDisplayHomeAsUpEnabled(true)
                actionBar.title = screenTitle
                // important for accessibility
                requireActivity().title = screenTitle
            }
        }
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.menu_site_creation, menu)
    }

    override fun onMenuItemSelected(menuItem: MenuItem) = when (menuItem.itemId) {
        R.id.help -> {
            onHelp()
            true
        }
        else -> false
    }

    companion object {
        const val EXTRA_SCREEN_TITLE = "extra_screen_title"
    }
}
